game-core/src/test/java/games/strategy/triplea/delegate/battle/MustFightBattleExecutablesTest.java

```diff
@@ -0,0 +1,1670 @@
+import games.strategy.engine.data.properties.GameProperties;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.Constants;
+import games.strategy.triplea.attachments.TechAttachment;
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.delegate.IExecutable;
+import java.math.BigDecimal;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Set;
+import junit.framework.AssertionFailedError;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.triplea.java.collections.IntegerMap;
+import org.triplea.sound.ISound;
+import org.triplea.util.Tuple;
+
+@ExtendWith(MockitoExtension.class)
+@SuppressWarnings("UnmatchedTest")
+class MustFightBattleExecutablesTest {
+
+  @Mock GameData gameData;
+  @Mock GameProperties gameProperties;
+  @Mock BattleTracker battleTracker;
+
+  @Mock Territory battleSite;
+  @Mock Territory retreatSite;
+  @Mock GamePlayer attacker;
+  @Mock GamePlayer defender;
+
+  @Mock Unit unit1;
+  @Mock UnitType unit1Type;
+  @Mock UnitAttachment unit1Attachment;
+
+  @Mock Unit unit2;
+  @Mock UnitType unit2Type;
+  @Mock UnitAttachment unit2Attachment;
+
+  enum BattleTerrain {
+    WATER,
+    LAND
+  }
+
+  private MustFightBattle newBattle(final BattleTerrain terrain) {
+    when(gameData.getProperties()).thenReturn(gameProperties);
+
+    final UnitCollection mockUnitCollection = mock(UnitCollection.class);
+    when(battleSite.getUnitCollection()).thenReturn(mockUnitCollection);
+
+    if (terrain == WATER) {
+      when(battleSite.isWater()).thenReturn(true);
+      final IntegerMap<GamePlayer> players = new IntegerMap<>();
+      players.add(defender, 1);
+      players.add(attacker, 1);
+      when(mockUnitCollection.getPlayerUnitCounts()).thenReturn(players);
+      when(gameProperties.get(SEA_BATTLE_ROUNDS, -1)).thenReturn(100);
+    } else {
+      when(battleSite.getOwner()).thenReturn(defender);
+      when(gameProperties.get(LAND_BATTLE_ROUNDS, -1)).thenReturn(100);
+    }
+
+    final RelationshipTracker mockRelationshipTracker = mock(RelationshipTracker.class);
+    when(gameData.getRelationshipTracker()).thenReturn(mockRelationshipTracker);
+    lenient().when(mockRelationshipTracker.isAtWar(attacker, defender)).thenReturn(true);
+    lenient().when(mockRelationshipTracker.isAllied(attacker, attacker)).thenReturn(true);
+    lenient().when(mockRelationshipTracker.isAllied(defender, defender)).thenReturn(true);
+    lenient().when(mockRelationshipTracker.isAllied(defender, attacker)).thenReturn(false);
+    lenient().when(mockRelationshipTracker.isAllied(attacker, defender)).thenReturn(false);
+
+    return new MustFightBattle(battleSite, attacker, gameData, battleTracker);
+  }
+
+  private void assertThatStepIsMissing(
```

Comment:
nit, we use 'depth-first-method-ordering'. I'm not sure if it's always the cleanest for tests. The ordering essentially is that methods should be defined as soon after first usage as possible. IE:
```
@Test
void foo() {
  helper1();
  helper2();
  // stuff
}

void helper1() { ... }

void helper2() { ... }

@Test
void foo2() {
   helper1();
   helper2();
  :
  :
}

```